lib/vnet/ipc_credentials.go

```diff
@@ -0,0 +1,207 @@
+	})
+	if err != nil {
+		return nil, trace.Wrap(err, "generating server TLS certificate")
+	}
+	clientCertPEM, err := clientCA.GenerateCertificate(tlsca.CertificateRequest{
+		PublicKey: clientSigner.Public(),
+		Subject:   pkix.Name{CommonName: "client"},
+		NotAfter:  now.Add(certTTL),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err, "generating client TLS certificate")
+	}
+
+	return &ipcCredentials{
+		server: credentials{
+			trustedCAPEM: clientCAPEM,
+			certPEM:      serverCertPEM,
+			signer:       serverSigner,
+		},
+		client: credentials{
+			trustedCAPEM: serverCAPEM,
+			certPEM:      clientCertPEM,
+			signer:       clientSigner,
+		},
+	}, nil
+}
+
+func (c *credentials) serverTLSConfig() (*tls.Config, error) {
+	caPool := x509.NewCertPool()
+	if !caPool.AppendCertsFromPEM(c.trustedCAPEM) {
+		return nil, trace.Errorf("Failed to parse CA")
+	}
+
+	tlsCert, err := keys.TLSCertificateForSigner(c.signer, c.certPEM)
+	if err != nil {
+		return nil, trace.Wrap(err, "parsing cert")
+	}
+
+	return &tls.Config{
+		Certificates: []tls.Certificate{tlsCert},
+		ClientCAs:    caPool,
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+		MinVersion:   tls.VersionTLS13,
+	}, nil
+}
+
+func (c *credentials) clientTLSConfig() (*tls.Config, error) {
+	caPool := x509.NewCertPool()
+	if !caPool.AppendCertsFromPEM(c.trustedCAPEM) {
+		return nil, trace.Errorf("Failed to parse CA")
+	}
+
+	tlsCert, err := keys.TLSCertificateForSigner(c.signer, c.certPEM)
+	if err != nil {
+		return nil, trace.Wrap(err, "parsing cert")
+	}
+
+	return &tls.Config{
+		Certificates: []tls.Certificate{tlsCert},
+		RootCAs:      caPool,
+		MinVersion:   tls.VersionTLS13,
+	}, nil
+}
+
+const (
+	caFileName   = "ca.pem"
+	certFileName = "cert.pem"
+	keyFileName  = "key.pem"
+)
+
+func (c *credentials) write(dir string) (err error) {
+	keyPEM, err := keys.MarshalPrivateKey(c.signer)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for fileName, data := range map[string][]byte{
+		caFileName:   c.trustedCAPEM,
+		certFileName: c.certPEM,
+		keyFileName:  keyPEM,
+	} {
+		filePath := filepath.Join(dir, fileName)
+		if err := os.WriteFile(filePath, data, 0600); err != nil {
+			return trace.Wrap(err, "writing service credential file %s", filePath)
+		}
+	}
+	return nil
+}
+
+func readCredentials(dir string) (*credentials, error) {
+	caBytes, err := os.ReadFile(filepath.Join(dir, caFileName))
+	if err != nil {
+		return nil, trace.Wrap(err, "reading service credential file %s", caFileName)
+	}
+	certBytes, err := os.ReadFile(filepath.Join(dir, certFileName))
+	if err != nil {
+		return nil, trace.Wrap(err, "reading service credential file %s", certFileName)
+	}
+	keyBytes, err := os.ReadFile(filepath.Join(dir, keyFileName))
+	if err != nil {
+		return nil, trace.Wrap(err, "reading service credential file %s", keyFileName)
+	}
+	signer, err := keys.ParsePrivateKey(keyBytes)
+	if err != nil {
+		return nil, trace.Wrap(err, "parsing service private key")
+	}
```

Comment:
Should we parse the other certificates here too, so we don't move forward with potentially invalid data? These are external input after all.

Reply:
imo it would be a bit redundant to parse the cert twice, we'll catch the error when we parse it eventually

Reply:
Fair enough.